Fixes #12898 - Adds content view and lifecycle environment to content…

johnpmitsch opened the pull request:
…_facet attr_accessible

This should allow the user to set the content_view_id and lifecycle_environment_id to a host (content_facet) but not other content_facet attributes like uuid

To test in the rails console
This should work:
```Host.new({:name => "hi", :content_facet_attributes => {:content_view_id => 1}})```

This should throw a mass assignment error:
```Host.new({:name => "hi", :content_facet_attributes => {:uuid => "hi"}})```

jlsherrill commented:
Would you be able to add a couple of simple tests to validate that some attributes can be updated while others can't?

johnpmitsch commented on app/models/katello/host/content_facet.rb:
```diff
@@ -18,6 +18,8 @@ class ContentFacet < Katello::Model
       validates :host, :presence => true, :allow_blank => false
       validates_with Validators::ContentViewEnvironmentValidator
 
+      attr_accessible :content_view, :content_view_id, :lifecycle_environment, :lifecycle_environment_id, :host
```
@jlsherrill updated to add the tests. I had to add both content_view and content_view_id (same with lifecycle_environment) here for passing tests in test/models/host/content_facet_test.rb but I think its fine to set all of these as accessible?

jlsherrill replied:
yeah i think thats fine, i wouldn't think the content_view and lifecycle_environment ones would be needed though?

johnpmitsch replied:
You would think so, but if I remove that [this test](https://github.com/Katello/katello/pull/5753/files#diff-bed4686568d79cce918e7c5566807cd5R18) fails because it uses :content_view and :lifecycle_environment as params. I can change the test, but I think we would want to be able to create a  content_facet by passing the whole content_view or lifecycle_environment

jlsherrill replied:
works for me!

daviddavis replied:
That's really an abuse of attr_accessible. Moreover, when we drop protected attributes, you won't able to set content_view and lifecycle_environment via mass assignment anyway.

johnpmitsch replied:
@daviddavis While foreman is using protected attributes, what do you think the best way to avoid mass assignment errors for these attributes would be?

daviddavis replied:
Three options:

1. Set the values explicitly on the model (e.g. https://gist.github.com/daviddavis/c89bf440cc2d763ccc43)
2. Use a ActionController::Parameters object and permit the keys you want (see http://api.rubyonrails.org/v4.2.4/classes/ActionController/Parameters.html#method-i-permit)
3. Use a factory

johnpmitsch commented:
@jlsherrill @daviddavis

 I've changed the [create test](https://git.io/vgsY6) to use lifecycle_environment_id and content_view_id. I kept these in attr_accessible, as I think that would work best with the [way we are passing the params](https://github.com/Katello/katello/blob/master/app/overrides/foreman/activation_keys/_host_environment_select.html.erb#L18) from the UI. 

Since we don't actually pass just content_view and lifecycle_environment (without _id) in the code afaik it wouldn't be worth testing IMO (I realize this kind of conflicts with what I said earlier, interested to hear if your opinions differ). 

In the dynflow task to destroy the host, I changed the [uuid to be set explicitly](https://git.io/vgsOu) so it won't throw any mass assignment errors

daviddavis commented:
:+1: from me on the attributes aspect.

johnpmitsch commented:
@jlsherrill @daviddavis updated the tests to set a content facet explicitly

jlsherrill commented:
@johnpmitsch seeing an error during registration:

ActiveModel::MassAssignmentSecurity::Error

Can't mass-assign protected attributes for Katello::Host::ContentFacet: uuid
- "/home/vagrant/git/katello/app/lib/actions/katello/host/register.rb:49:in `finalize'"

jlsherrill commented on app/lib/actions/katello/host/register.rb:
```diff
@@ -42,8 +42,9 @@ def humanized_name
-          host.subscription_facet.update_attributes(:uuid => input[:uuid])
+          host.content_facet.uuid = input[:uuid]
+          host.subscription_facet.uuid = input[:uuid]
+          host.save!
```
i think you need to do a save on each facet

johnpmitsch replied:
thanks updated to save on each facet

jlsherrill commented on test/models/host/content_facet_test.rb:
```diff
@@ -14,12 +15,22 @@ class ContentFacetBase < ActiveSupport::TestCase
+      assert ::Host.new(:name => "foo", :content_facet_attributes => { :content_view_id => view.id, :lifecycle_environment_id => environment.id })
+    end
+
+    def test__protected_content_facet_attributes
```
two _'s in this method name

jlsherrill commented:
Looks like the current test failures are just these two tests:  https://github.com/Katello/katello/blob/master/test/models/concerns/host_managed_extensions_test.rb#L81-L91

changing them to use the content_facet seems to fix them

johnpmitsch commented on test/models/concerns/host_managed_extensions_test.rb:
```diff
@@ -80,14 +78,14 @@ def test_update_does_not_update_content_host
 
     def test_update_with_invalid_cv_env_combo
-      host = FactoryGirl.create(:host, :with_content, :content_view => @library_view, :lifecycle_environment => @library)
+      host = FactoryGirl.create(:host)
```
@jlsherrill I updated to just create the host and then add the content view and lifecycle environment

jlsherrill replied:
i don't think this will fix the problem, but we'll see!

jlsherrill commented on test/models/concerns/host_managed_extensions_test.rb:
```diff
@@ -79,18 +77,18 @@ def test_update_does_not_update_content_host
-      host.content_view = @library_view
-      host.lifecycle_environment = @library
-      assert host.save!
+      host = FactoryGirl.create(:host, :with_content, :content_view => view, :lifecycle_environment =>  library)
```
I think this is the reason it failed, library isn't defined here, and neither is view.

jlsherrill commented:
ACK
